### PR TITLE
Viite 82 managing road addresses after updated mtk imports

### PR DESCRIPTION
**Pull request by hestrompa**
Includes tasks :
VIITE-474 VVH Client to fetch change information for Viite (regular + complementary)
VIITE-475 Mapping, cases 1+2: Combined links
VIITE-481 Batch process to go with floating links check
VIITE-528 More test cases

### CODE REVIEW

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala [hunk 1]
```diff
@@ -145,6 +154,52 @@ class RoadAddressService(roadLinkService: RoadLinkService, eventbus: DigiroadEve
...
+  def resolveChanges(roadlinks: Seq[RoadLink], changedRoadLinks: Seq[ChangeInfo], addresses: Map[Long, Seq[RoadAddress]]): Map[Long, Seq[RoadAddress]] = {
+    withDynTransaction {
+
+      val newRoadAddresses = RoadAddressChangeInfoMapper.resolveChangesToMap(addresses, roadlinks, changedRoadLinks)
+
+      val savedRoadAddresses = newRoadAddresses.mapValues(_.map(r =>
+        if (r.id == NewRoadAddress) {
+          val roadWithNewGeom = r.copy(id = RoadAddressDAO.getNextRoadAddressId, geom = GeometryUtils.truncateGeometry3D(roadlinks.filter(link => link.linkId == r.linkId).head.geometry, r.startMValue, r.endMValue))
+          RoadAddressDAO.create(Seq(roadWithNewGeom))
+          println("New road address created> linkId: "+roadWithNewGeom.linkId+" id:"+roadWithNewGeom.id)
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala]: Replace with a logger entry

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala [hunk 2]
```diff
@@ -145,6 +154,52 @@ class RoadAddressService(roadLinkService: RoadLinkService, eventbus: DigiroadEve
...
+  }
+
+  def resolveChanges(roadlinks: Seq[RoadLink], changedRoadLinks: Seq[ChangeInfo], addresses: Map[Long, Seq[RoadAddress]]): Map[Long, Seq[RoadAddress]] = {
+    withDynTransaction {
+
+      val newRoadAddresses = RoadAddressChangeInfoMapper.resolveChangesToMap(addresses, roadlinks, changedRoadLinks)
+
+      val savedRoadAddresses = newRoadAddresses.mapValues(_.map(r =>
+        if (r.id == NewRoadAddress) {
+          val roadWithNewGeom = r.copy(id = RoadAddressDAO.getNextRoadAddressId, geom = GeometryUtils.truncateGeometry3D(roadlinks.filter(link => link.linkId == r.linkId).head.geometry, r.startMValue, r.endMValue))
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala]: Create a map of road links from id -> road link

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala [hunk 3]
```diff
@@ -144,6 +153,38 @@ class RoadAddressService(roadLinkService: RoadLinkService, eventbus: DigiroadEve
...
+
+      val removedIds = addresses.values.flatten.filterNot(a =>
+        returnAddressMap.getOrElse(a.linkId, Seq()).exists(_.id == a.id)).map(_.id).toSet
+      removedIds.grouped(500).foreach(s => {RoadAddressDAO.expireById(s)
+        logger.debug("Expired: "+s.mkString(","))
+      })
+
+      val changedRoadParts = addressesToCreate.map(a => a.roadNumber -> a.roadPartNumber).groupBy(_._1).mapValues(seq => seq.map(_._2).toSet)
+
+      changedRoadParts.foreach { case (road, roadParts) => roadParts.foreach(part => recalculateRoadAddresses(road, part)) }
```
Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala]: reload addresses if recalculation happened for any of those

File: digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala [hunk 4]
```diff
@@ -110,7 +117,9 @@ class RoadAddressService(roadLinkService: RoadLinkService, eventbus: DigiroadEve
     //pure linkId based queries, maybe something related to the zoomLevel we are in map level.
     val fetchMissingRoadAddressStartTime = System.currentTimeMillis()
     val (floatingViiteRoadLinks, addresses, floating) = Await.result(fetchRoadAddressesByBoundingBoxF, Duration.Inf)
-    val missingLinkIds = linkIds -- floating.keySet -- addresses.keySet
+    val complementedWithChangeAddresses = applyChanges(complementedRoadLinks, filteredChangedRoadLinks, addresses)
+    //TODO sub complementedWithChangeAddresses to addresses
```
Comment (oappicgi) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala]: this was for later case, right?

Comment (tvenhola) [inline digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/RoadAddressService.scala]: We create the list of missing link ids by taking the list of all link ids, removing the link ids that have floating road addresses, non-floating addresses, and post-transfer addresses.

This method should be refactored when speed improvements are developed in Sprint 2.

**Pull request merged**